### PR TITLE
Expand forking to editor

### PR DESCRIPTION
**Pull request by donfrigo**
### Description of the Change

Added functionality to fork projects when they are modified. As per the requirements stated in Superblocks 2.0 PR. 

### Benefits

Users can now easily save their own version of projects quickly and intuitively without having to click 'fork'.

### Verification Process

Move, delete, create or rename files in 'foreign' projects to see that at that point a new own project is created. ~~A new fork is also created when renaming the project.~~

### CODE REVIEW

File: src/epics/explorer/deleteItem.epic.ts [hunk 1]
```diff
@@ -14,30 +14,58 @@
...
-        );
+                .pipe(
+                    switchMap(() => {
+                        return [panesActions.closePane(action.data.id), explorerActions.deleteItemSuccess(action.data.id)];
+                    }),
+                    catchError(() => [explorerActions.deleteItemFail(action.data.id)])
+                );
+        } else {
+            // fork with new tree structure
+            return projectService.createProject({
```
Comment (ostap1010) [inline src/epics/explorer/deleteItem.epic.ts]: This part of code seems to be repeated all over in every action. Please refactor it out in separate epic and just produce some action to trigger that epic if possible.

File: src/epics/explorer/createItem.epic.ts [hunk 2]
```diff
@@ -16,29 +16,55 @@
...
+                    name: project.name,
+                    description: project.description,
+                    files: state$.value.explorer.tree
+                }).pipe(
+                    switchMap(() => [explorerActions.createItemSuccess()]),
+                    catchError(() => [ explorerActions.createItemFail(explorerState.itemNameValidation.itemId) ])
+                );
+            } else {
+                // fork with new tree structure
+                return projectService.createProject({
```
Comment (ostap1010) [inline src/epics/explorer/createItem.epic.ts]: This part of code seems to be repeated all over in every action. Please refactor it out in separate epic and just produce some action to trigger that epic if possible.

File: src/actions/projects.actions.ts [hunk 3]
```diff
@@ -141,6 +142,13 @@ export const projectsActions = {
             data: { newName }
        };
     },
+    RENAME_PROJECT_FAIL: 'RENAME_PROJECT_FAIL',
```
Comment (ostap1010) [inline src/actions/projects.actions.ts]: This action does not seem to be used, please delete it for now

File: src/epics/projects/renameProject.epic.ts [hunk 4]
```diff
@@ -14,27 +14,37 @@
...
+                name,
+                description,
+                files
+            })
+                .pipe(
+                    switchMap(() => projectService.getProjectById(project.id)),
+                    map(projectsActions.updateProjectSuccess),
+                );
+        } else {
+            throw Error('You can only rename your own project!');
```
Comment (ostap1010) [inline src/epics/projects/renameProject.epic.ts]: Error can be omitted cause UI is not exposing this possibility

File: src/services/project.service.ts [hunk 5]
```diff
@@ -57,7 +57,12 @@ export const projectService = {
         return fetchJSON(process.env.REACT_APP_PROJECT_API_BASE_URL + '/projects/' + id, {
             method: 'PUT',
             body: data
-        });
+        }).pipe(
+            switchMap(r => (r.ok ? r.statusText : throwError(r.statusText))),
+            catchError(err => {
+                throw err;
```
Comment (ostap1010) [inline src/services/project.service.ts]: Looks like some debug code remaining. Please clean up.

Comment (donfrigo) [reply src/services/project.service.ts]: removed

File: src/epics/projects/renameProject.epic.ts (hunk 4 above)
Comment (donfrigo) [reply src/epics/projects/renameProject.epic.ts]: fixed

**Pull request merged**